File: algo_basic_python_java/s9.-sorting/5.kth_largest_element/5_kth_largest_element.py
class Solution:
    """
    @param n: An integer
    @param nums: An array
    @return: the Kth largest element
    """
    def kthLargestElement(self, n, nums):
        # write your code here
        if not nums:
            return -1
        return self.quick_select(nums, 0, len(nums) - 1, n)
    
    def quick_select(self, nums, start, end, n):
        if start == end:
            return nums[start]
        
        left = start
        right = end
        pivot = nums[start + (end - start)//2]

        while left <= right:
            while left <= right and nums[left] > pivot:
                left+=1
            while left <= right and nums[right] < pivot:
                right-=1
            if left <= right:
                nums[left], nums[right] = nums[right], nums[left]
                left+=1
                right-=1
        
        #WARNING!
        # if k = 1th, then start == j, also k start from 1
        if start + n - 1 <= right:
            return self.quick_select(nums, start, right, n)
        if start + n - 1 >= left:
            #WARNING!
            # since left part don't care, k should reduct them
            return self.quick_select(nums, left, end, n - (left - start ))
        return nums[right + 1]

File: lintcode_algo_basic_python_java/s9.-sorting/5.kth_largest_element/5_kth_largest_element.java
public class Solution {
    /**
     * @param n: An integer
     * @param nums: An array
     * @return: the Kth largest element
     */
    public int kthLargestElement(int n, int[] nums) {
        // write your code here
        if (nums == null || nums.length ==0) {
            return -1;
        }

        return quickSelect(nums, 0, nums.length - 1, n);
    }

    private int quickSelect(int[] nums, int start, int end, int n) {
        if (start >= end) {
            return nums[start];
        }
        
        int mid = start + (end - start)/2;
        int pivot = nums[mid];
        int left = start, right = end;
        
        while (left <= right) {
            // keep traversing
            while (left <= right && nums[left] > pivot) {
                left++;
            }
            while (left <= right && nums[right] < pivot) {
                right--;
            }
            if (left <= right) {
                // swap value
                int tmp = nums[left];
                nums[left] = nums[right];
                nums[right] = tmp;
                left++;
                right--;
            }
        }
        // locate kth element
        if (start + n - 1 <= right) {
            return quickSelect(nums, start, right, n);
        } 
        if (start + n - 1 >= left) {
            return quickSelect(nums, left, end, n - (left - start));
        }
        return nums[right + 1];

    }        
}